File: java/leetcode/src/Task10.java
import org.junit.jupiter.api.Test;

public class Task10 {
    public boolean isMatch(String s, String p) {
        //condition

        //define
        int m = s.length();
        int n = p.length();

        boolean[][] dp = new boolean[m+1][n+1];

        //init
        dp[0][0] = true;
        for (int i = 2; i <= n; i++) {
            if(p.charAt(i-1) == '*' && p.charAt(i-2) != '*'){
                dp[0][i] = dp[0][i-2];
            }
        }

        //loop
        for (int i = 1; i <= m ; i++) {
            for (int j = 1; j <= n ; j++) {
                if(s.charAt(i-1) == p.charAt(j-1) || p.charAt(j-1) == '.'){
                    dp[i][j] = dp[i-1][j-1];
                }else if(p.charAt(j-1) == '*' && j-2 >= 0){
                    if(s.charAt(i-1) == p.charAt(j-2) || p.charAt(j-2) == '.'){
                        dp[i][j] = dp[i-1][j] || dp[i][j-2];
                    }else{
                        dp[i][j] = dp[i][j-2];
                    }
                }
            }
        }

        //return
        return dp[m][n];

    }

    @Test
    void test(){
        String s = "aaa";
        String p = "ab*a*c*a";
        System.out.println(isMatch(s,p));
    }
}

File: py/Task10.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if not p:
            return not s

        m = len(s)
        n = len(p)

        # define
        dp = [[False] * (n+1) for _ in range(m + 1)]

        # init
        dp[0][0] = True
        for i in range(2, n+1):
            if p[i - 1] == '*':
                dp[0][i] = dp[0][i-2]

        # loop
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                ss = s[i - 1]
                pp = p[j - 1]
                if pp == '.' or ss == pp:
                    dp[i][j] = dp[i - 1][j - 1]
                elif pp == '*':
                    dp[i][j] = dp[i][j - 2]
                    ppp = p[j - 2]
                    if ppp == '.' or ss == ppp:
                        dp[i][j] = dp[i][j] or dp[i - 1][j]
        return dp[-1][-1]


s = ""
p = ".*"
obj = Solution()
print(obj.isMatch(s, p))
